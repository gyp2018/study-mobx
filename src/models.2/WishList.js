import { types } from 'mobx-state-tree'

export const WishListItem = types
  .model({
      price: types.number,
      name: types.string,
    image: types.optional(types.string, '')
  })
  .actions(self => ({
    changeName(newName) {
      self.name = newName;
    },
    changePrice(newPrice) {
      self.price = newPrice
    },
    changeImage(newImage) {
      self.image = newImage
    }
  }))

export const WishList = types
  .model({
    items: types.optional(types.array(WishListItem), [])
  })
  .actions(self => ({
    add(item) {
      self.items.push(item)
    }
  }))
