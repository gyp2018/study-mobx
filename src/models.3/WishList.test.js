import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree'
import { WishListItem, WishList } from './WishList'

it('can create a instance of model', () => {
  const item = WishListItem.create({
    'name': 'PlayStation 4 Pro',
    'price': 499.00,
  })

  expect(item.price).toBe(499.00)
  expect(item.image).toBe('')

  item.changeName('XBox One')
  expect(item.name).toBe('XBox One')
})

it('can creat a wishist', () => {
  const list = WishList.create({
    items: [{
      'name': 'PlayStation 4 Pro',
      'price': 499.00,
    }]
  })

  expect(list.items.length).toBe(1)
  expect(list.items[0].price).toBe(499.00)
})

it('can add new items', () => {
  const list = WishList.create()
  const states = []
  onSnapshot(list, snapshot => {
    states.push(snapshot)
  })

  list.add({
    'name': 'XBox One',
    'price': 399.00,
  })

  expect(list.items.length).toBe(1)
  expect(list.items[0].name).toBe('XBox One')
  
  list.items[0].changeName('Nintendo Switch')
  expect(list.items[0].name).toBe('Nintendo Switch')

  expect(getSnapshot(list)).toMatchSnapshot()
  expect(states).toMatchSnapshot()
})

it('can add new items - 2', () => {
  const list = WishList.create()
  const patches = []
  onPatch(list, patch => {
    patches.push(patch)
  })

  list.add({
    'name': 'XBox One',
    'price': 399.00,
  })

  list.items[0].changeName('Nintendo Switch')

  expect(patches).toMatchSnapshot()
})
