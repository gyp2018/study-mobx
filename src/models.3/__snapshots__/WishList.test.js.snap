// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items - 2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "XBox One",
      "price": 399,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Nintendo Switch",
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Nintendo Switch",
      "price": 399,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "XBox One",
        "price": 399,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Nintendo Switch",
        "price": 399,
      },
    ],
  },
]
`;
